infra/base-images/base-builder/jcc/jcc.go

```diff
@@ -165,15 +166,72 @@ func CorrectMissingHeaders(bin string, cmd []string) (bool, error) {
 	return false, nil
 }
 
-func compile(bin string, args []string) (int, string, string) {
-	cmd := exec.Command(bin, args...)
+func RunCommand(cmd *exec.Cmd) (int, string, string) {
+	// Executes a command and returns the output.
 	var outb, errb bytes.Buffer
 	cmd.Stdout = &outb
 	cmd.Stderr = &errb
 	cmd.Run()
 	return cmd.ProcessState.ExitCode(), outb.String(), errb.String()
 }
 
+func ExecOriginalCommand(bin string, args []string) (int, string, string) {
+	// Executes the original command.
+	cmd := exec.Command(bin, args...)
+	return RunCommand(cmd)
+}
+
+func Contains(slice []string, item string) bool {
+	// Checks if the slice contains item.
+	for _, s := range slice {
+		if strings.EqualFold(s, item) {
+			return true
+		}
+	}
+	return false
+}
+
+func FindTargetFile(args []string) string {
+	// Finds the fuzz target file by file extension.
+	suffixes := []string{".cpp", ".cc", ".cxx", ".c++", ".c"}
+	for _, arg := range args {
+		if Contains(suffixes, filepath.Ext(arg)) {
+			return filepath.Base(arg)
+		}
+	}
+	// Uses a time stamp as AST filename if no target file is found.
+	return time.Now().Format("20060102_150405.000")
+}
+
+func RemoveIfEmpty(filepath string) {
+	// Removes filepath if it is empty.
+	info, _ := os.Stat(filepath)
+	if info.Size() == 0 { os.Remove(filepath) }
+}
+
+func GenerateAST(bin string, args []string) (int, string, string) {
+	// Generates AST.
+	newArgs := append(args, "-Xclang", "-ast-dump=json", "-fsyntax-only")
+
+	targetFile := FindTargetFile(args)
+	filePath := filepath.Join("/out", fmt.Sprintf("%s.ast", targetFile))
+
+	cmdStr := fmt.Sprintf("%s %s > %s", bin, strings.Join(newArgs, " "), filePath)
+	cmd := exec.Command("sh", "-c", cmdStr)
+	retCode, stdout, stderr := RunCommand(cmd)
+
+	RemoveIfEmpty(filePath)
```

Comment:
Can you add a comment for why this happens? And why we want to delete if empty?

Reply:
The main reason I found is the command does not include any C/CPP files.
For example, `ada-url` first builds `parse.o` with `parser.c`, which generates the AST we need, then it builds `parse` with `parse.o`, which generates an empty AST file.
I will add a comment about this.

These empty files are not useful so I cleaned them up.
I suppose a better solution is avoiding generating them in the first place (e.g., do not use our wrapper when there are no C/CPP files), but I think this is the simplest way to avoid side effects.

Reply:
I see. Please add a comment here to explain this :)

Reply:
Done!